modules/transformation.py

```diff
@@ -0,0 +1,609 @@
+from apache_beam.io import ReadFromPubSub, WriteToBigQuery, BigQueryDisposition
+import apache_beam.io.fileio as fileio
+from apache_beam.io.gcp.bigquery_tools import FileFormat, RetryStrategy
+from dateutil import parser as date_parser
+
+from modules import constants, utils
+
+
+class ReadHarFiles(beam.PTransform):
+    def __init__(self, subscription=None, _input=None):
+        super().__init__()
+        self.subscription = subscription
+        self.input = _input
+
+    def expand(self, p):
+        # PubSub pipeline
+        if self.subscription:
+            files = (
+                p
+                | ReadFromPubSub(subscription=self.subscription, with_attributes=True)
+                | beam.Filter(lambda e: e.attributes["objectId"].endswith(".har.gz"))
+                | "GetFileName"
+                >> beam.Map(
+                    lambda e: f"gs://{e.attributes['bucketId']}/{e.attributes['objectId']}"
+                )
+                | beam.io.ReadAllFromText(with_filename=True)
+            )
+        # GCS pipeline
+        else:
+            matching = (
+                self.input if ".har.gz" in self.input else f"{self.input}/*.har.gz"
+            )
+
+            # TODO using ReadAllFromText instead of ReadFromText[WithFilename] to avoid listing file sizes locally
+            #   https://stackoverflow.com/questions/60874942/avoid-recomputing-size-of-all-cloud-storage-files-in-beam-python-sdk
+            #   https://issues.apache.org/jira/browse/BEAM-9620
+            #   not an issue for the java SDK
+            #     https://beam.apache.org/releases/javadoc/2.37.0/org/apache/beam/sdk/io/contextualtextio/ContextualTextIO.Read.html#withHintMatchesManyFiles--
+            # files = p | beam.io.ReadFromTextWithFilename(matching)
+
+            files = (
+                p
+                # TODO replace with match continuously for streaming?
+                | fileio.MatchFiles(matching)
+                | "ExtractPath" >> beam.Map(lambda f: f.path)
+                | beam.Reshuffle()
+                | beam.io.ReadAllFromText(with_filename=True)
+            )
+
+        return files
+
+
+class WriteBigQuery(beam.PTransform):
+    def __init__(self, table, schema, streaming=None):
+        super().__init__()
+        self.table = table
+        self.schema = schema
+        self.streaming = streaming
+
+    def resolve_params(self):
+        if self.streaming:
+            # streaming pipeline
+            return {
+                "create_disposition": BigQueryDisposition.CREATE_IF_NEEDED,
+                "write_disposition": BigQueryDisposition.WRITE_APPEND,
+                "with_auto_sharding": True,
+                # parameters for STREAMING_INSERTS
+                "method": WriteToBigQuery.Method.STREAMING_INSERTS,
+                "ignore_unknown_columns": True,
+                "insert_retry_strategy": RetryStrategy.RETRY_ON_TRANSIENT_ERROR,
+                # parameters for FILE_LOADS
```

Comment:
Can these be removed?

Reply:
I will refactor this to put these options behind a pipeline parameter so the user can decide which insert method to use at runtime.